charms/jupyter-ui/tests/integration/test_charm_ambient.py

```diff
@@ -0,0 +1,197 @@
+CONFIG = yaml.safe_load(Path("./config.yaml").read_text())
+APP_NAME = "jupyter-ui"
+JUPYTER_IMAGES_CONFIG = "jupyter-images"
+VSCODE_IMAGES_CONFIG = "vscode-images"
+RSTUDIO_IMAGES_CONFIG = "rstudio-images"
+PORT = CONFIG["options"]["port"]["default"]
+HEADERS = {
+    "kubeflow-userid": "",
+}
+
+AFFINITY_OPTIONS = [
+    {
+        "configKey": "test-affinity-config-1",
+        "displayName": "Test Affinity Config-1",
+        "affinity": dict(
+            nodeAffinity=dict(
+                requiredDuringSchedulingIgnoredDuringExecution=[
+                    dict(
+                        matchExpressions=[
+                            dict(key="lifecycle", operator="In", values=["kubeflow-notebook-1"])
+                        ]
+                    )
+                ]
+            )
+        ),
+    },
+]
+
+TOLERATIONS_OPTIONS = [
+    {
+        "groupKey": "test-tolerations-group-1",
+        "displayName": "Test Tolerations Group 1",
+        "tolerations": [
+            dict(effect="NoSchedule", key="dedicated", operator="Equal", value="big-machine")
+        ],
+    },
+]
+DEFAULT_PODDEFAULTS = [
+    "poddefault1",
+    "poddefault2",
+]
+
+
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, request):
+    """Build and deploy the charm.
+
+    Assert on the unit status.
+    """
+    # Keep the option to run the integration tests locally
+    # by building the charm and then deploying
+    entity_url = (
+        await ops_test.build_charm("./")
+        if not (entity_url := request.config.getoption("--charm-path"))
+        else entity_url
+    )
+    image_path = METADATA["resources"]["oci-image"]["upstream-source"]
+    resources = {"oci-image": image_path}
+
+    await ops_test.model.deploy(
+        entity_url, resources=resources, application_name=APP_NAME, trust=True
+    )
+
+    # NOTE: idle_period is used to ensure all resources are deployed
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=60 * 10, idle_period=30
+    )
+    assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+
+    # Deploying grafana-agent-k8s and add all relations
+    await deploy_and_assert_grafana_agent(
+        ops_test.model, APP_NAME, metrics=False, dashboard=False, logging=True
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_and_relate_dependencies(ops_test: OpsTest):
+    """Deploy and integrate Istio dependencies with the application under test."""
+    await deploy_and_integrate_service_mesh_charms(
+        app=APP_NAME,
+        model=ops_test.model,
+        channel="2/edge",
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_ui_is_accessible(ops_test: OpsTest):
+    """Verify that UI is accessible through the ingress gateway."""
+    await assert_path_reachable_through_ingress(
+        http_path="/jupyter/",
```

Comment:
suggestion (non-blocking): can we hoist it with a constant, for developers to find it at a glance when maintaining it?